django/contrib/auth/migrations/0010_update_proxy_permissions.py

```diff
@@ -0,0 +1,57 @@
+from django.contrib.auth import get_permission_codename
+from django.db import migrations
+
+
+def update_proxy_model_permissions(apps, schema_editor):
+    Permission = apps.get_model('auth', 'Permission')
+    ContentType = apps.get_model('contenttypes', 'ContentType')
+
+    for Model in apps.get_models():
+        opts = Model._meta
+
+        if not opts.proxy:
+            continue
+
+        concrete_content_type = ContentType.objects.get_for_model(Model,
+                                                                  for_concrete_model=True)
+
+        default_proxy_codenames = [
+            get_permission_codename(action, opts)
+            for action in opts.default_permissions
+        ]
+        proxy_custom_codenames = {codename for codename, __ in opts.permissions}
+
+        concrete_custom_codenames = {
+            codename for codename, __ in opts.concrete_model._meta.permissions
+        }
+
+        custom_codenames = list(concrete_custom_codenames - proxy_custom_codenames)
+
+        codenames = default_proxy_codenames + custom_codenames
+
+        if opts.app_label == opts.concrete_model._meta.app_label:
```

Comment:
This is the safe part, where we move the existing permissions to the new content type.